hist/hist/inc/THnChain.h

```diff
@@ -0,0 +1,266 @@
+hs.GetAxis(1)->SetRangeUser(0, 0.1); // select a subrange
+TH1* h0 = hs.Projection(0);
+\endcode
+
+*/
+
+class THnChain : public TObject
+{
+ public:
+   /// Default constructor.
+   ///
+   /// \param name name of the histogram to work on
+   THnChain(const char* name) : fName(name) { fFiles.SetOwner(kTRUE); }
+
+   /// Add a new file to this chain.
+   ///
+   /// \param fileName path of the file to add
+   void AddFile(const char* fileName)
+   {
+      fFiles.Add(new TObjString(fileName));
+
+      // Initialize axes from first seen instance.
+      if (fAxes.GetEntries() == 0) {
+         THnBase* hs = ReadHistogram(fileName);
+
+         if (hs) {
+            const Int_t naxes = hs->GetNdimensions();
+            for (Int_t i = 0; i < naxes; ++i) {
+               fAxes[i] = hs->GetAxis(i);
+            }
+         }
+      }
+   }
+
+   /// Get an axis from the histogram.
+   ///
+   /// \param i index of the axis to retrieve
+   ///
+   /// This function requires that a file containing the histogram was
+   /// already added with `AddFile`.
+   ///
+   /// Properties set on the axis returned by `GetAxis` are propagated to all
+   /// histograms in the chain, so it can e.g. be used to set ranges for
+   /// projections.
+   TAxis* GetAxis(Int_t i)
+   {
+      if (i < 0 || i >= fAxes.GetEntries()) {
+         return nullptr;
+      }
+
+      return static_cast<TAxis*>(fAxes[i]);
+   }
+
+   TH1* Projection(Int_t xDim, Option_t* option = "")
+   {
+      // Forwards to `THnBase::Projection()`.
+      Int_t dim[1] = {xDim};
+      return static_cast<TH1*>(ProjectionAny(1, dim, option));
+   }
+
+   TH2* Projection(Int_t yDim, Int_t xDim, Option_t* option = "")
+   {
+      // Forwards to `THnBase::Projection()`.
+      Int_t dim[2] = {xDim, yDim};
+      return static_cast<TH2*>(ProjectionAny(2, dim, option));
+   }
+
+   TH3* Projection(Int_t xDim, Int_t yDim, Int_t zDim, Option_t* option = "")
+   {
+      // Forwards to `THnBase::Projection()`.
+      Int_t dim[3] = {xDim, yDim, zDim};
+      return static_cast<TH3*>(ProjectionAny(3, dim, option));
+   }
+
+   THnBase* ProjectionND(Int_t ndim, const Int_t* dim, Option_t* option = "")
+   {
+      // Forwards to `THnBase::ProjectionND()`.
+      return static_cast<THnBase*>(ProjectionAny(ndim, dim, option));
+   }
+
+ private:
+   TString fName; ///< name of the histogram
+
+   TObjArray fFiles; ///< a list of files to extract the histogram from
+   TObjArray fAxes;  ///< the list of histogram axes
+
+   /// Projects all histograms in the chain.
+   ///
+   /// See `THnBase::Projection` for parameters and their semantics.
+   TObject* ProjectionAny(Int_t ndim, const Int_t* dim, Option_t* option = "")
+   {
+      if (ndim == 0) {
+         return nullptr;
+      }
+
+      TObject* h_merged = nullptr;
+      for (const auto& obj : fFiles) {
+         THnBase* hs = ReadHistogram(static_cast<TObjString*>(obj)->String());
+
+         if (!hs || !CheckConsistency(*hs, fAxes)) {
```

Comment:
Can you provide separate error messages to the user whether the hist is found of inconsistent?